lib/smack/SmackRep.cpp

```diff
@@ -623,10 +751,38 @@ string SmackRep::getPrelude() {
   s << endl;
   s << "// Memory region declarations";
   s << ": " << memoryRegions.size() << endl;
-  for (unsigned i=0; i<memoryRegions.size(); ++i)
-    s << "var " << memReg(i) << ": " << memType(i) << ";" << endl;
+  for (unsigned i=0; i<memoryRegions.size(); ++i) {
+    for (unsigned j = 0; j < (SmackOptions::InferFieldOverlap? 4 : 1); j++) {
+      unsigned size = 8 << j;
+      s << "var " << memPath(i, size) 
+	<< ": " << memType(i, size) 
+	<< ";" << endl;
+    }
+  }
+  s << endl;
+
+  s << "// Type declarations" << endl;
+  for (unsigned i = 8; i <= width; i <<= 1) {
```

Comment:
This loop should not be bounded by the width of pointers. It should be bounded by the maximum width of integers.